### DISCUSSION TITLE
关于分段。

### DISCUSSION THREAD

**Discussion by liusy58** [liusy58/WhyThis#7] [Q&A] [answered]
最近看到一些说法说分段是历史的糟粕，但是我不这样认为，大家可以说说自己的理解，我认为分段在link阶段有十分重要的运用。

**Comment by xfoxfu** [comment, marked as answer]
我觉得我们现在的链接器设计成这个样子主要是因为以前的分段内存的设计。是因为有了段（segment），所以程序里也就出现了节（section）。我觉得段本身的设计是没有问题的，但是 x86 ISA 中为段引入了偏移，是为了解决 16 位寄存器下的 20 位寻址问题，这个技术的引入我认为是 ISA 走的偏路（虽然也可以理解为特定历史时期下芯片晶体管数量的妥协）。

通过引入段偏移，实际上程序的地址空间被人为地割裂开，程序的指针因为其指向的内容不同而产生了不同的含义，这一特性的引入使得程序的调试、操作系统的设计和编译器的开发都变得更为复杂了。在拥有 32 位乃至 64 位地址空间，以及大量 GPR 的当下，段偏移的价值非常有限。

\* 本评论从[原 issue](https://github.com/liusy58/WhyThis/issues/4#issuecomment-821245962) 复制。

**Comment by Inokinoki** [comment]
感觉大家的回答（可能问题本身也）限制于 386 兼容 pc 的设计了，同期其实还有非常多其他的处理器。

section 其实挺有必要的，比如哈佛架构里数据代码分离的设计，在 ELF 里具体实现就可以在运行时将 .text section 放入代码储存器，将 .rodata .bss 放入数据储存器。

**Reply by liusy58** [reply]
又收获了一些奇奇怪怪的知识，好的，这周我们争取一起整理一下这个问题。

**Reply by liusy58** [reply]
因为我对分段的理解也不深刻，所以有一些局限，老哥可以畅所欲言的，这没关系，只要有道理我们都可以写上来。

**Reply by liusy58** [reply]
因为暂时只想到了链接时候，所以就往那个方向展开了。
